setup.py

```diff
@@ -0,0 +1,149 @@
+try:
+    from sphinx.setup_command import BuildDoc
+    SPHINX_AVAILABLE = True
+except ImportError:
+    pass
+
+here = abspath(dirname(__file__))
+
+
+def read(*parts):
+    with codecs.open(join(here, *parts), 'r') as fp:
+        return fp.read()
+
+
+def find_version(*file_paths):
+    version_file = read(*file_paths)
+    version_match = re.search(r"^__version__ = ['\"]([^'\"]*)['\"]",
+                              version_file, re.M)
+    if version_match:
+        return version_match.group(1)
+    raise RuntimeError("Unable to find version string.")
+
+
+cmdclass = {}
+if SPHINX_AVAILABLE:
+
+    class MakeDocs(BuildDoc):
+        """Port of old `cylc-make-docs`. Removed in #2989.
+
+        This class extends the Sphinx command class for setuptools. With
+        a difference that it tries to mimic the behaviour of the previous
+        `cylc-make-docs`.
+
+        So first it will execute `make-commands.sh`, which builds the
+        commands help information, in the appendices.
+
+        Then, instead of calling one builder, this class will call the
+        builder fot he single HTML, and also the builder for the multiple
+        HTML documentation.
+
+        Finally, one more tweak in this class is to move the doctrees
+        folder (in the same level as the documentation) to within the
+        documentation folder, named `.doctrees`, as before with
+        `cylc-make-docs`.
+        """
+
+        def run(self):  # type: () -> None
+            try:
+                self.spawn(["./doc/src/custom/make-commands.sh"])
+            except DistutilsExecError as e:
+                self.warn("Failed to run make-commands.sh")
+                raise e
+            self.do_run("html", "built-sphinx")
+            self.do_run("singlehtml", "built-sphinx-single")
+
+        def do_run(self, builder: str, output_dir: str):
+            """
+            Args:
+                builder (str): name of the Sphinx builder
+                output_dir (str): directory to write the documentation produced
+            """
+            self.builder = builder
+            self.builder_target_dirs = [
+                (builder, join(self.build_dir, output_dir))]
+            super().run()
+            # move doctrees to $build_dir/.doctrees
+            correct_doctrees = join(self.builder_target_dirs[0][1],
+                                    ".doctrees")
+            rmtree(correct_doctrees, ignore_errors=True)
+            move(self.doctree_dir, correct_doctrees)
+
+    cmdclass["build_sphinx"] = MakeDocs
+
+
+install_requires = [
+    'colorama==0.4.1',
+    'isodatetime==1!2.0.0',
+    'jinja2==2.10',
+    'markupsafe==1.1.1',
+    'python-jose==3.0.1',
+    'pyzmq==18.0.1'
+]
+tests_require = [
+    'codecov==2.0.15',
+    'coverage==4.5.3',
+    'pytest-cov==2.6.1',
+    'pytest==4.4.0',
+    'pycodestyle==2.5.0',
+    'testfixtures==6.6.2'
+]
+
+extra_requires = {
+    'empy': ['EmPy==3.3.4'],
+    'docs': ['sphinx==2.0.0'],
```

Comment:
This isn't true! The Cylc docs actually builds just fine with ancient Sphinx (1.2 I think).

I'd rather not invent an arbitrary version dependency like this as it could make Cylc harder to install by creating conflicts with other modules.

Reply:
Sure thing. I was lazy and simply did `pip install sphinx`, then grabbed the version I was using. Will give it a try tomorrow with sphinx 1.2.

Note that this is only for tests, and also that we are using the sphinx integration with setuptools, which may force us to move to a more recent version. Though I agree, if it conflicts with any other library necessary for running Cylc, sphinx has lower priority

Reply:
BTW: min Sphinx version will go to what we use for Rose (1.5.6 I think) when we copy across the Cylc Tutorial.

Reply:
On it now, see comments below

>This isn't true! The Cylc docs actually builds just fine with ancient Sphinx (1.2 I think).

The last 1.2 release dates from `Release 1.2.3 (released Sep 1, 2014)`. So we can probably use a newer version.

>I'd rather not invent an arbitrary version dependency like this as it could make Cylc harder to install by creating conflicts with other modules.

Had a look at sphinx list of dependencies, and looks like we won't have any conflicts if their dependencies? (**NB: this is an optional dependency, only available when building from sources. When you install via `pip install cylc`, sphinx dependency will not be installed.**).

```
install_requires = [
    'sphinxcontrib-applehelp',
    'sphinxcontrib-devhelp',
    'sphinxcontrib-jsmath',
    'sphinxcontrib-htmlhelp',
    'sphinxcontrib-serializinghtml',
    'sphinxcontrib-qthelp',
    'Jinja2>=2.3',
    'Pygments>=2.0',
    'docutils>=0.12',
    'snowballstemmer>=1.1',
    'babel>=1.3,!=2.0',
    'alabaster>=0.7,<0.8',
    'imagesize',
    'requests>=2.5.0',
    'setuptools',
    'packaging',
]
```

Their Jinja version is compatible with ours. And looking at their 1.5.6 dependency list we have:

```
    'six>=1.5',
    'Jinja2>=2.3',
    'Pygments>=2.0',
    'docutils>=0.11',
    'snowballstemmer>=1.1',
    'babel>=1.3,!=2.0',
    'alabaster>=0.7,<0.8',
    'imagesize',
    'requests>=2.0.0',
```

Sphinx 1.5.6 forces users to install babel 1.3 (last release before 2.0), which was released July 29th 2013. Which I think could be a problem in some sites due to security.

Babel releases after 1.3 include fixes that could be useful for us too

```
Version 2.6.0
* Python 3.6 invalid escape sequence deprecation fixes (@scop) (PR #528)

Version 2.0
- Improved Python 3 support.
```

So I think it would be best to use the latest version of Sphinx, and keep up with their updates if possible?

>BTW: min Sphinx version will go to what we use for Rose (1.5.6 I think) when we copy across the Cylc Tutorial.

1.5.6 is about two years old `Release 1.5.6 (released May 15, 2017)`.

Just up to `Release 1.7.9 (released Sep 05, 2018)`:

23 deprecated
318 bugs fixed
15 features removed
42 incompatible changes

Not counting new features. Out of the bug fixes, 43 fixes crashes in Sphinx:

```
    #6136: :name: option for math directive causes a crash
    #5958: versionadded directive causes crash with Python 3.5.0
    #5659: linkcheck: crashes for a hyperlink containing multibyte character
    #5800: todo: crashed if todo is defined in TextElement
    #5490: latex: enumerated list causes a crash with recommonmark
    #5493: gettext: crashed with broken template
    #5563: latex: footnote_references generated by extension causes a LaTeX builder crash
    #5437: autodoc: crashed on modules importing eggs
    #5399: Sphinx crashes if unknown po file exists
    #5361: crashed on incremental build if document uses include directive
    #5320: intersphinx: crashed if invalid url given
    #5326: manpage: crashed when invalid docname is specified as man_pages
    #5291: autodoc crashed by ForwardRef types
    #5016: crashed when recommonmark.AutoStrictify is enabled
    #5022: latex: crashed with docutils package provided by Debian/Ubuntu
    #5048: crashed with numbered toctree
    #5019: autodoc: crashed by Form Feed Character
    #5143: autodoc: crashed on inspecting dict like object which does not support sorting
    #5161: html: crashes if copying static files are failed
    #3329: i18n: crashed by auto-symbol footnote references
    latex: deeply nested enumerated list which is beginning with non-1 causes LaTeX engine crashed
    #4919: node.asdom() crashes if toctree has :numbered: option
    #4931: autodoc: crashed when handler for autodoc-skip-member raises an error
    #4931: autodoc: crashed when subclass of mocked class are processed by napoleon module
    #5007: sphinx-build crashes when error log contains a “%” character
    #4885, #4887: domains: Crashed with duplicated objects
    #4783: On windows, Sphinx crashed when drives of srcdir and outdir are different
    #4878: Crashed with extension which returns invalid metadata
    autodoc: crashed when invalid options given
    #2286: Sphinx crashes when error is happens in rendering HTML pages
    #4662: any refs with term targets crash when an ambiguity is encountered
    #4279: Sphinx crashes with pickling error when run with multiple processes and remote image
    #4096: C++, don’t crash when using the wrong role type. Thanks to mitya57.
    #4070, #4111: crashes when the warning message contains format strings (again)
    #4152: HTML writer crashes if a field list is placed on top of the document
    #4063: Sphinx crashes when labeling directive .. todolist::
    #4006: Fix crash on parallel build
    #4070: crashes when the warning message contains format strings
    #3754: HTML builder crashes if HTML theme appends own stylesheets
    #3758: Sphinx crashed if logs are emitted in conf.py
    #3796: env.resolve_references() crashes when non-document node given
    #3803: Sphinx crashes with invalid PO files
    #3661: sphinx-build crashes on parallel build
```

So I'm +1 for keeping this change for now, unless we are not able to build Cylc documentation with the latest Sphinx. Hope that makes sense :+1:

Reply:
Makes sense to me ... what do you think @oliver-sanders ?

Reply:
I don't think we should demand such a specific install base when there is no actual dependency on these versions.

Take a look at the Sphinx requirements or other projects like [notebook](https://github.com/jupyter/notebook/blob/master/setup.py#L99), [jupyterhub](https://github.com/jupyterhub/jupyterhub/blob/master/requirements.txt), no dependency on specific versions and some dependencies specified without versions at all. I think this is the normal approach.

My inclination would be to specify what's actually *required* and let the package manager do the rest, something like this:

```diff
 install_requires = [
-    'colorama==0.4.*',
+    'colorama',  # no specific version requirement
-    'isodatetime==1!2.0.*',
+    'isodatetime>=1!2.0.*',  # py3 requirement
     'jinja2==2.10',  # hard-code templating engine requirements
-    'markupsafe==1.1.*',
+    'markupsafe>=1.0.0',  # no specific version requirement could remove the 1.0.0?
-    'python-jose==3.0.*',
+    'python-jose>=3.0.0',  # for security this makes sense
-    'pyzmq==18.0.*'
+    'pyzmq>=18.0.0'  # for security this makes sense
 ]
 tests_require = [
-    'codecov==2.0.*',
+    'codecov',  # no specific version requirement
-    'coverage==4.5.*',
+    'coverage',  # no specific version requirement
-    'pytest-cov==2.6.*',
+    'pytest-cov',  # no specific version requirement
-    'pytest==4.4.*',
+    'pytest',  # no specific version requirement
-    'pycodestyle==2.5.*',
+    'pycodestyle',  # no specific version requirement
-    'testfixtures==6.6.*'
+    'testfixtures'  # no specific version requirement
 ]

 extra_requires = {
    'empy': ['EmPy==3.3.*'],  # hard-code templating engine requirements
-    'docs': ['sphinx==2.0.*'],
+    'docs': ['sphinx'],  # no specific version requirement (yet)
     'all': []
 }
```

Reply:
>My inclination would be to specify what's actually required and let the package manager do the rest, something like this:

I think that's a bad idea... version freezing is quite common in other projects too, like [Flask](https://github.com/pallets/flask/blob/86bf9dca72cd188813e26ef5dd972101ef8983cf/setup.py#L36), [requests](https://github.com/kennethreitz/requests/blob/9742da7f91f595796a412a8150aacc00191be039/setup.py#L45), [pandas](https://github.com/pandas-dev/pandas/blob/0610a60d06c190cdd597480962f98b97de7b4286/setup.py#L36), etc.

And [this blog](https://blog.miguelgrinberg.com/post/the-package-dependency-blues) tries to give an example of where maintaining a project without frozen dependencies can be harder. We will be basically saying that if users have the oldest version of `colorama` installed, that requirement has been fulfilled. Meaning that if they have a bug, they might just report that something is not working.

And for JupyterHub/pyzmq/notebooks, etc. I believe the author/maintainers simply started the project that way. Then started freezing the dependencies as required. e.g. from when [they froze](https://github.com/jupyterhub/jupyterhub/issues/350) sqlalchemy after a user reported an error. The error reported in this case, is quite similar to the example given in the blog above. User with old dependency, maintainers did not specify version, bug is created.

If we are being too strict, perhaps we can specify at least a minimum. But going without any version makes it easier to get users reporting issues due to version incompatibilities.

Reply:
I think Matt chimed in on Riot to suggest a sort of compromise: leave dependency versions open (ish) on master, and then freeze them later for released versions.

Here's another article that insists on "pinning"/freezing versions: https://nvie.com/posts/pin-your-packages/ - although just for releases, I think.

Reply:
A question for @oliver-sanders (or others) - what exactly is the downside of being unnecessarily specific (e.g. requiring a recent version of Sphinx when an older version actually still works OK for us)?  Is it that we increase the chance of version conflicts that pip can't solve?  e.g. cylc requires version 1.2 of some package, but some other installed program requires version 1.5 of the same package, and pip can only support one version of it (within a single Python environment).   

If that's the case, we'll presumably still have the same problem with releases - even though (as argued above) releases really need to be be frozen.

So ... all python programs really need to be installed with their own private virtual environments, to avoid this problem?

Reply:
Anyhow, we don't have to make a definitive decision now - we could stick with what @kinow has done for the moment, and change later if necessary right?  (it would be good to get this PR merged soon!)

Reply:
> So ... all python programs really need to be installed with their own private virtual environments,

Or go all the way and containerize everything ...

Reply:
I actually had a problem with version conflicts in cylc-uiserver when a version point was set for [Jinja2](https://github.com/cylc/cylc-uiserver/issues/28) with `jinja2==2.9.6`, and kind of agree with @oliver-sanders   ... It can cause issues in the order of install at the least..  Perhaps setting a low bar and having an open ended interval would do `>=`

Reply:
Not all users may have control over the version of software installed on their system (and have no experience with virtual environments)...
Even JupyterHub has no frozen [requirements](https://github.com/jupyterhub/jupyterhub/blob/master/requirements.txt)

Reply:
>I actually had a problem with version conflicts in cylc-uiserver when a version point in one of the dependencies was set for Jinja2 with jinja2==2.9.6, and kind of agree with @oliver-sanders ... It can cause issues in the order of install at the least.. Perhaps setting a low bar and having an open ended interval would do >=

Yup. I agreed on that one too. And the version has been changed now to `'jinja2>=2.10.1, <2.11.0',`.

>Not all users may have control over the version of software installed on their system (and have no experience with virtual environments)...

Agreed. But suppose a user in that situation tells us about a bug in his environment, and we forget to check the version of his dependencies. We could spend a long time chasing down a bug that is actually caused by a version of a dependency that we do not support (could be a method that does not exist, etc).

See [this blog](https://blog.miguelgrinberg.com/post/the-package-dependency-blues) I mentioned earlier to @oliver-sanders too, about a cautionary tale for this.

And while Cylc never asked users to install specific version of libraries... it was only because we provided with batteries included. So they always got the right version of `isodatetime`, `jinja2`, etc, because we were bundling these dependencies. That could the reason why we do not have a history of issues with dependency management in Cylc.

>Even JupyterHub has no frozen requirements

Yup, not frozen, but they do specify ranges. What we have now in `setup.py` is not very different from what they have.

The difference with JupyterHub is that they are less strict about semver. They use simply `requests` instead of specifying which version of `requests`. But if you have the very first release of requests in your python path, there is no guarantee that a) it will work, and b) that it does not have any security bugs.

And if you look at [`requests`](https://github.com/kennethreitz/requests/blob/d22d219d34c81a1f5e599a43b3beb9449c5cf15e/setup.py#L44-L58), they are pretty detailed about what they allow in their dependency range.

Said all that, once merged, we can discuss and change these values. The discussion about the versions of our dependencies in `setup.py` has been the main blocker for this pull request I think. So for me, I would not mind removing the version ranges completely here, if that means we can merge it now :slightly_smiling_face:

Reply:
> A question for @oliver-sanders (or others) - what exactly is the downside of being unnecessarily specific

A belated answer from @oliver-sanders...

* Requiring a recent version of software is fine, it might cause grumbles but ultimately it forces people to upgrade.
* Pinning to a specific version e.g. `program==1.2.*`, however, has downsides:
    * When `1.3.0` is released our users cannot upgrade.
        * A lot of software projects have very rapid release cycles (Sphinx is now excepting PRs for 3.0.0).
        * For us the lifetime of a Cylc deployment is ~1yr.
    * Users may have to downgrade software to install Cylc.
        * I actually see this fairly often, Conda gives you a little message when it happens.
        * As the dependency base grows this could cause a deadlock with other modules. This is quite likely as I can foresee users wanting one environment containing Cylc along with all of their scientific, visualisation and number crunching goodness.

**If others feel strongly I'm happy to pipe down**. Otherwise my compromise would be change the `==` to `>=` i.e. specify open ended ranges where sensible.

If I were doing it I would make the non-functional dependencies (i.e. documentation and testing) much more lenient as these are for developers not end users.

Reply:
These are all good arguments. I still few safer locking on the patch version. The counter argument is that a fast-paced project could release a new version that breaks Cylc for users. Which I think is as bad as having to downgrade dependencies.

But right now we have no users depending on this version, so we can try your approach first and fallback in case a user reports issues with dependencies. Feel free to submit a pull request against `master` to fix these dependencies. We can start with more lenient approach, and change it later if we have issue with the dependencies. (i.e. `>=` for main dependencies and remove for non-functional dependencies :+1:)